var/spack/repos/builtin/packages/suite-sparse/package.py

```diff
@@ -236,13 +241,25 @@ def install(self, spec, prefix):
                 # *_LIBRARIES is critical to pick up static
                 # libraries (if intended) and also to avoid
                 # unintentional system blas/lapack packages
-                + " -DBLAS_LIBRARIES=%s" % spec["blas"].libs
-                + " -DLAPACK_LIBRARIES=%s" % spec["lapack"].libs
+                + ' -DBLAS_LIBRARIES="%s"' % (";".join(spec["blas"].libs))
+                + ' -DLAPACK_LIBRARIES="%s"' % (";".join(spec["lapack"].libs))
+                + " -DENABLE_CUDA=%s" % ("ON" if "+cuda" in spec else "OFF")
+                + " -DSUITESPARSE_USE_OPENMP=%s" % ("ON" if "+openmp" in spec else "OFF")
+                # Older versions use the negative flag NOPENMP: "OFF" means use
+                # OpenMP:
+                + " -DNOPENMP=%s" % ("OFF" if "+openmp" in spec else "ON")
+                + " -DCMAKE_VERBOSE_MAKEFILE=ON"
             ]
 
         if spec.satisfies("%gcc platform=darwin"):
             make_args += ["LDLIBS=-lm"]
 
+        if "%cce" in spec:
+            # Assume the proper Cray CCE module (cce) is loaded:
+            craylibs_path = env["CRAYLIBS_" + machine().upper()]
+            env.setdefault("LDFLAGS", "")
+            env["LDFLAGS"] += " -Wl,-rpath," + craylibs_path
```

Comment:
Unfortunately, `spec.target` is not good -- e.g. it gives `zen3` and `machine()` is `x86_64` -- I need something that returns `x86_64` or `aarch64` since the `cce` module defines `CRAYLIBS_X86_64` and `CRAYLIBS_AARCH64`.

If the environment variable is not defined, we get an exception, right? I though this is good enough. What guard do you have in mind?

I've added this logic (in previous PRs) in a few packages where I have needed this tweak before -- `strumpack`, `butterflypack`, and in `mfem` itself. Without explicitly adding this `rpath`, what happens is that the `cce` compiler links the libraries of `suite-sparse` with some libraries in the path from `CRAYLIBS_*` but does not add that path to the `rpath` (when linking a library). When linking an executable `cce` adds automatically the path from `CRAYLIBS_*` to the `rpath` and links again with the same libraries in that path, so there is no issue when the final linking is done by `cce`. However, when building `+rocm`, `mfem` compiles and links everything with `hipcc`, so the path from `CRAYLIBS_*` needs to be added explicitly as `rpath` and linking should also link with the needed `cce` libraries. The `mfem` package does that here: https://github.com/spack/spack/blob/b5920de26a55a64531cccd6c6f146902b5e599a4/var/spack/repos/builtin/packages/mfem/package.py#L987-L1001

Unfortunately, `hipcc` adds the `rpath` as `RUNPATH` and not as `RPATH` and thus the `rpath` from `CRAYLIBS_*` is not considered when loading libraries needed by mfem executables (examples or miniapps) like the `suite-sparse` libraries. This leads to "library not found" errors at runtime.

Maybe one option that I can try in the `mfem` package is to add the `CRAYLIBS_*` path while explicitly disabling the new dtags with `--disable-new-dtags` -- I think this will tell `hipcc` to add the `rpath` arguments as `RPATH` which should propagate to dependencies like `suite-sparse`. If this works, I can probably remove the logic about `CRAYLIBS_*` from other packages -- I'm not sure if that will break anything outside of `mfem` though, e.g. the `xsdk-examples`.

Reply:
> Unfortunately, spec.target is not good -- e.g. it gives zen3 and machine() is x86_64 -- I need something that returns x86_64 or aarch64

`str(spec.target.family)` will do that for you

> If the environment variable is not defined, we get an exception, right? I though this is good enough. What guard do you have in mind?

I was thinking it might be good to avoid a key error if `CRAYLIBS_*` isn't in `env` so like:

```
craylibs_path = env.get("CRAYLIBS_" + machine().upper(), None)
# Maybe this should fail if craylibs_path is None, but we can throw a clearer exception than keyerror
# like "this variable needs to be defined and isn't"
```

> When linking an executable cce adds automatically the path from CRAYLIBS_* to the rpath and links again with the same libraries in that path, so there is no issue when the final linking is done by cce. However, when building +rocm, mfem compiles and links everything with hipcc, so the path from CRAYLIBS_* needs to be added explicitly as rpath 

Thanks for that explanation: I think it makes sense to me why we wouldn't have auto-added an rpath argument

> Maybe one option that I can try in the mfem package is to add the CRAYLIBS_* path while explicitly disabling the new dtags with --disable-new-dtags -- I think this will tell hipcc to add the rpath arguments as RPATH which should propagate to dependencies like suite-sparse. 

I'd be curious if that works (that would help me think on how to improve this in general), but I think they way you have it now is preferable.

I'm not clear how this would propagate to dependencies: generally changes applied to a parent won't propagate (although there are ways to do that). FWIW the spack compiler wrappers generally add `--disable-new-dtags` automatically. If we could get the `hip` compiler to use the spack wrappers, this should happen "for free" (but I don't have a notion of whether that would be easy, or have unintended side effects).

Reply:
Addressed the handling of the `CRAYLIBS_*` variable in f02270302d.

Reply:
I tried commenting out the `if "%cce" in spec` section in `suite-sparse` and building with testing `mfem +rocm +suite-sparse` without modifying the mfem package (in particular without adding `--disable-new-dtags` in the `cce` handling). It turns out that actually works now -- I think because all the implicit Cray libraries that `cce` links with are now also added by mfem here: https://github.com/spack/spack/blob/b5920de26a55a64531cccd6c6f146902b5e599a4/var/spack/repos/builtin/packages/mfem/package.py#L987-L1001

I think, this did not work before because I did not have all the needed Cray libraries, e.g. I added `libpgas-shmem` later. I suspect that if these libs are explicitly linked with when building an executable with proper `RPATH` or `RUNPATH` then they are found as dependencies of the executable and if needed as dependencies-of-dependencies, they they don't need to be re-resolved.

On the other hand, however, when the `if "%cce" in spec` section in `suite-sparse` is commented out, the `suite-sparse` libraries are not self contained in the sense that `ldd libumfpack.so` shows "not found" for the Cray libraries that live in the `CRAYLIBS_*` path. For this reason, I think it's better to keep adding the extra `rpath` when building with `cce`.